### PR TITLE
CORE-12034 Make `DigitalSignature` an interface on public API

### PR DESCRIPTION
**Pull request by kyriathar**
- adds `DigitalSignature.WithKeyId` implementation and updates use sites
- adds tests asserting AMQP serializability for types `DigitalSignatureWithKey` and `DigitalSignatureWithKeyId`

api PR: [#1011](https://github.com/corda/corda-api/pull/1011)

### CODE REVIEW

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt [hunk 1]
```diff
@@ -1,9 +1,10 @@
 package net.corda.crypto.core
 
+import net.corda.v5.base.types.OpaqueBytes
 import net.corda.v5.crypto.DigitalSignature
 import java.security.PublicKey
 
 class DigitalSignatureWithKey(
```
Comment (mnesbit) [inline libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt:7]: Needs equality and hash

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package net.corda.crypto.core
+
+import net.corda.v5.base.types.OpaqueBytes
+import net.corda.v5.crypto.DigitalSignature
+import net.corda.v5.crypto.SecureHash
+
+class DigitalSignatureWithKeyId(
```
Comment (mnesbit) [inline libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt:7]: Needs equality and hash

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    private val by: SecureHash,
+    bytes: ByteArray
+    ) : DigitalSignature.WithKeyId, OpaqueBytes(bytes) {
+    // TODO Not sure why Kotlin generated getter cannot override?
+    override fun getBy() = by
```
Comment (mnesbit) [inline libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt:11]: When doing Kotlin -> Java properties have to become explicit getters and setters.

Comment (mnesbit) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt:11]: I am worried that this class hides a lot of details in the innocent inheritance from OpaqueBytes. It isn't clear that this has value equality and hashcode and it isn't clear that measures are taken to make the getBytes() read-only/a copy

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt (hunk 1 above)
Comment (mnesbit) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt:7]: Does this concrete class have to be marked as @CordaSerializable to be AMQP serializable? Not sure we can inherit that from the interface?

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt (hunk 3 above)
Comment (mnesbit) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt:11]: We need to verify that this is AMQP serializable and doesn't require an AMQP custom serializer. I currently only see Kryo custom serializers

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt (hunk 1 above)
Comment (kyriathar) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKey.kt:7]: Added tests showing they are AMQP serializable without `CordaSerializable` so seems they do inherit it from interface.

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt (hunk 3 above)
Comment (kyriathar) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/DigitalSignatureWithKeyId.kt:11]: > I am worried that this class hides a lot of details in the innocent inheritance from OpaqueBytes. It isn't clear that this has value equality and hashcode

Yes indeed, they are offered through `ByteSequence` for all byte wrapper types extending it.

**Review by driessamyn — DISMISSED** [review]
LGTM

**Review by driessamyn — APPROVED** [review]
LGTM

**Review by mnesbit — APPROVED** [review]
OK

**Pull request merged**